res/res_pjsip/redirect.c

```diff
@@ -0,0 +1,403 @@
+			endpoint->redirect_method == AST_SIP_REDIRECT_URI_CORE ? "uri_core" : "unknown");
+		return 0;
+	}
+
+	return 1;
+}
+
+/*!
+ * \internal
+ * \brief Check if a URI has already been visited (loop detection)
+ */
+static int is_uri_visited(struct ast_sip_redirect_context *context, const char *uri)
+{
+	struct visited_uri *visited;
+
+	AST_LIST_TRAVERSE(&context->visited_uris, visited, list) {
+		if (!strcmp(visited->uri, uri)) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+/*!
+ * \internal
+ * \brief Add a URI to the visited list
+ */
+static int add_visited_uri(struct ast_sip_redirect_context *context, const char *uri)
+{
+	struct visited_uri *visited;
+
+	visited = ast_calloc(1, sizeof(*visited));
+	if (!visited) {
+		return -1;
+	}
+
+	ast_copy_string(visited->uri, uri, sizeof(visited->uri));
+	AST_LIST_INSERT_TAIL(&context->visited_uris, visited, list);
+
+	return 0;
+}
+
+/*!
+ * \internal
+ * \brief Extract q-value from a Contact header
+ *
+ * \param contact The Contact header
+ * \return The q-value (default 1.0 if not present or invalid)
+ */
+static float extract_q_value(pjsip_contact_hdr *contact)
+{
+	pjsip_param *param;
+	static const pj_str_t Q_STR = { "q", 1 };
+
+	/* Search for q parameter in the contact header */
+	param = pjsip_param_find(&contact->other_param, &Q_STR);
+	if (!param) {
+		/* No q parameter, use default */
+		return 1.0f;
+	}
+
+	/* Parse the q value */
+	if (param->value.slen > 0) {
+		char q_buf[16];
+		char *endptr;
+		float q_val;
+		int len = param->value.slen < sizeof(q_buf) - 1 ? param->value.slen : sizeof(q_buf) - 1;
+
+		memcpy(q_buf, param->value.ptr, len);
+		q_buf[len] = '\0';
+
+		errno = 0;
+		q_val = strtof(q_buf, &endptr);
+		if (errno == 0 && endptr != q_buf && q_val >= 0.0f && q_val <= 1.0f) {
+			return q_val;
+		}
+	}
+
+	/* Invalid q value, use default */
+	return 1.0f;
+}
+
+/*!
+ * \internal
+ * \brief Insert a contact into the sorted list by q-value (highest first)
+ *
+ * \param list The list to insert into
+ * \param new_contact The contact to insert
+ */
+static void insert_contact_sorted(struct redirect_contact_list *list, struct redirect_contact *new_contact)
+{
+	struct redirect_contact *contact;
+
+	/* Find the insertion point - contacts with higher q values come first */
+	AST_LIST_TRAVERSE_SAFE_BEGIN(list, contact, list) {
+		if (new_contact->q_value > contact->q_value) {
+			/* Insert before this contact */
+			AST_LIST_INSERT_BEFORE_CURRENT(new_contact, list);
+			return;
+		}
+	}
+	AST_LIST_TRAVERSE_SAFE_END;
+
+	/* If we get here, insert at the end */
+	AST_LIST_INSERT_TAIL(list, new_contact, list);
+}
+
+/*!
+ * \internal
+ * \brief Parse all Contact headers from a 3xx response and create a sorted list
+ *
+ * \param rdata The redirect response data
+ * \param contacts List to populate with parsed contacts
+ * \return Number of valid contacts found
+ */
+static int parse_redirect_contacts(pjsip_rx_data *rdata, struct redirect_contact_list *contacts)
```

Comment:
Pass in the state/context pointer, and use the endpoint to add name to the logging

I'm kind of a broken record lately with logging, but giving information on which endpoint is involved can be very valuable if someone goes back to try to understand what happened on a well used system

The same applies for the code going forward after this comment, I'm not going to comment on each one but I urge you to go back to each and try to make them more specific/detailed including endpoint name